### PR TITLE
Build: Enable React Fast Refresh for block development

### PR DESCRIPTION
**Pull request by gziolo**
Trac ticket: https://core.trac.wordpress.org/ticket/51750

Depends on #2263.

Brings the same functionality introduced in the Gutenberg plugin with https://github.com/WordPress/gutenberg/pull/28273. In effect, it brings React Fast Refresh support to WordPress core for block development with `@wordpress/scripts`.

The way I test it:

```bash
cd src/wp-content/plugins
npx @wordpress/create-block gutenpride -t @wordpress/create-block-tutorial-template
cd my-block
npm start -- --hot
```

Ensure your WordPress develop instance has `SCRIPT_DEBUG` set to `true`. It should be the default when using `npm run env:start`. You also need to activate the plugin with the Gurenpride block.

## Screenshots

https://user-images.githubusercontent.com/699132/143570765-0e1f0f25-21c8-4ded-8223-1756103b7895.mov

https://user-images.githubusercontent.com/699132/143570743-1c6a57d3-69e3-44d7-a073-051fc6d0c744.mov

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/script-loader.php [hunk 1]
```diff
@@ -214,6 +214,39 @@ function wp_get_script_polyfill( $scripts, $tests ) {
...
+		return;
+	}
+
+	$development_scripts = array(
+		'react-refresh-entry',
+		'react-refresh-runtime',
+	);
+
+	foreach ( $development_scripts as $script_name ) {
+		$assets = include ABSPATH . WPINC . "/js/dist/development/$script_name.php";
```
Comment (anton-vlasenko) [inline src/wp-includes/script-loader.php]: I'd add a check if `$assets` is an actual array.
If the file doesn't exist, `include` will return false, failing the function.

Comment (anton-vlasenko) [inline src/wp-includes/script-loader.php]: Please consider this a nitpick, but for consistency, I'd either use the concatenation operator or curly braces to concatenate strings across this function.

```php
$assets = include ABSPATH . WPINC . '/js/dist/development/' . $script_name . '.php';
```
or
```php
$assets = include ABSPATH . WPINC . "/js/dist/development/${script_name}.php";
```
Performance-wise, the first option is better.
It's a bit hard to read:
```php
"/wp-includes/js/dist/development/$script_name.js",
```

File: src/wp-includes/script-loader.php [hunk 2]
```diff
@@ -214,6 +214,39 @@ function wp_get_script_polyfill( $scripts, $tests ) {
...
+		$scripts->add(
+			'wp-' . $script_name,
```
Comment (anton-vlasenko) [inline src/wp-includes/script-loader.php:245]: The concatenation operator (`.`) is being used here (which is fine).
But, should the other places in this function have to be updated to use the concatenation operator too (for consistency and readability)?

File: src/wp-includes/script-loader.php [hunk 3]
```diff
@@ -214,6 +214,39 @@ function wp_get_script_polyfill( $scripts, $tests ) {
...
+			"/wp-includes/js/dist/development/$script_name.js",
```
Comment (anton-vlasenko) [inline src/wp-includes/script-loader.php]: Same as [my comment](https://github.com/WordPress/wordpress-develop/pull/2519#discussion_r844403706) above.

**Review by anton-vlasenko — CHANGES_REQUESTED** [review]
@gziolo It would be nice to add a PHPUnit test for `wp_register_development_scripts`.
I'd be happy to assist with that.
Also, please see my notes below.
Thank you.

File: src/wp-includes/script-loader.php (hunk 1 above)
Comment (gziolo) [reply src/wp-includes/script-loader.php]: Good point. I will check how it’s handled in other places.

File: src/wp-includes/script-loader.php (hunk 2 above)
Comment (gziolo) [reply src/wp-includes/script-loader.php:245]: Done 👍🏻

File: src/wp-includes/script-loader.php (hunk 1 above)
Comment (gziolo) [reply src/wp-includes/script-loader.php]: Added 👍🏻

**Comment by walbo** [comment]
Tested and works as expected (tested in firefox and chrome with wp-scripts). 🎉

**Review by anton-vlasenko — APPROVED** [review]
I haven't reviewed JavaScript changes, but the PHP part of the PR looks good to me.

**Comment by gziolo** [comment]
Committed with https://core.trac.wordpress.org/changeset/53140.

**Pull request closed without merging**